### PR TITLE
[Polly] Data flow reduction detection to cover more cases

### PR DESCRIPTION
**Pull request by kartcq**
The base concept is same as existing reduction algorithm where we get the list of candidate pairs <store,load>. But the existing algorithm works only if there is single binary operation between the load and store.
Example sum += a[i];

This algorithm extends to work with more than single binary operation as well. It is implemented using data flow reduction detection on basic block level. We propagate the loads, the number of times the load is used(flows into instruction) and binary operation performed until we reach a store.

Example sum += a[i] + b[i];
```
sum(Ld)     a[i](Ld)
      \  +  /
        tmp    b[i](Ld)
           \ + /
            sum(St)
```

In the above case the candidate pairs are formed by associating sum with all of its load inputs which are sum, a[i] and b[i]. Then check functions are used to filter a valid reduction pair ie {sum,sum}.

### CODE REVIEW

**Comment by kartcq** [comment]
ping @efriedma-quic @Meinersbur

File: polly/lib/Analysis/ScopBuilder.cpp [hunk 1]
```diff
@@ -2568,49 +2579,196 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
...
+  DenseMap<MemAccPair, MemoryAccess::ReductionType> ValidCandidates;
+  DominatorTree *DT = Stmt.getParent()->getDT();
+
+  // Iterate over all write memory accesses and check the loads flowing into
+  // it for reduction candidate pairs.
+  for (MemoryAccess *WriteMA : Stmt.MemAccs) {
+    if (WriteMA->isRead())
+      continue;
+    StoreInst *St = dyn_cast<StoreInst>(WriteMA->getAccessInstruction());
+    if (!St || St->isVolatile())
```
Comment (Meinersbur) [inline polly/lib/Analysis/ScopBuilder.cpp:2724]: Shouldn't volatile stores lead to invalidating the loads as other non-valid reductions? ScopDetection should already reject volatile access anyway, so this could be a `assert(!St->isVolatile())` or `assert(St->isSimple())`

File: polly/lib/Analysis/ScopBuilder.cpp [hunk 2]
```diff
@@ -2568,49 +2579,196 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
...
 
-    // If no overlapping access was found we mark the load and store as
-    // reduction like.
-    LoadMA->markAsReductionLike(RT);
-    StoreMA->markAsReductionLike(RT);
+      MemoryAccess::ReductionType RT = MaInFlowSetElem.second.second;
+      unsigned NumAllowableInFlow = 1;
+
+      // We allow the load to flow in exactly once for binary reductions
+      Valid = (MaInFlowSetElem.second.first == NumAllowableInFlow);
```
Comment (Meinersbur) [inline polly/lib/Analysis/ScopBuilder.cpp:2744]:
```suggestion
      bool Valid = (MaInFlowSetElem.second.first == NumAllowableInFlow);
```

File: polly/lib/Analysis/ScopBuilder.cpp [hunk 3]
```diff
@@ -2568,49 +2579,196 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
...
+  // it for reduction candidate pairs.
+  for (MemoryAccess *WriteMA : Stmt.MemAccs) {
+    if (WriteMA->isRead())
+      continue;
+    StoreInst *St = dyn_cast<StoreInst>(WriteMA->getAccessInstruction());
+    if (!St || St->isVolatile())
+      continue;
+
+    FlowInSetTy &MaInFlowSet = State[WriteMA->getAccessInstruction()];
+    bool Valid = false;
```
Comment (Meinersbur) [inline polly/lib/Analysis/ScopBuilder.cpp:2728]:
```suggestion
```
Only used inside the loop

**Review by Meinersbur — CHANGES_REQUESTED** [review]
Thanks for the patch. I really appreciate the good code comments.

Could you add a test that checks two independent reductions in the same loop, e.g.:
```
for (int i = 0; i < n; ++i) {
  *sum += A[i];
  *prod += B[i];
}
```
`ScopBuilder::buildEqivClassBlockStmts` should create two ScopStmts for the same BasicBlock here.

File: polly/lib/Analysis/ScopBuilder.cpp [hunk 4]
```diff
@@ -2568,49 +2579,196 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
...
+  //     (e.g., A[i] = B[i] * 3 and  A[i] = A[i] * A[i] + A[i])
+  //   o are used by a non binary operator or one which is not commutative
+  //     and associative (e.g., A[i] = A[i] % 3)
+  //   o might change the control flow            (e.g., if (A[i]))
+  //   o are used in indirect memory accesses     (e.g., A[B[i]])
+  //   o are used outside the current basic block
+  SmallPtrSet<const Instruction *, 8> InvalidLoads;
+
+  // Run the data flow analysis for all values in the basic block
+  for (Instruction &Inst : *Block) {
```
[the remark this answers is not in the document]
Comment (kartcq) [reply polly/lib/Analysis/ScopBuilder.cpp]: I tried to use Stmt.getInstructions() here. But for scop statements with multiple blocks, the getInstructions() API returns the instructions of first block only. I am not sure if it bug with the function or expected behavior.

[the remark this answers is not in the document]
Comment (Meinersbur) [reply polly/lib/Analysis/ScopBuilder.cpp]: This is expected behavior. `getInstructions()` cannot capture the entire control flow in a RegionStmt so it only represents the instructions of the header block. See https://github.com/llvm/llvm-project/blob/e76b257483e6c6743de0fa6eca4d0cc60e08385d/polly/lib/CodeGen/BlockGenerators.cpp#L444-L449 for rationale and how to handle.

File: polly/lib/Analysis/ScopBuilder.cpp [hunk 5]
```diff
@@ -2568,49 +2579,196 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
...
 }
 
-void ScopBuilder::checkForReductions(ScopStmt &Stmt) {
-  SmallVector<MemoryAccess *, 2> Loads;
-  SmallVector<std::pair<MemoryAccess *, MemoryAccess *>, 4> Candidates;
+/// Perform a data flow analysis on the current basic block to propagate the
+/// uses of loaded values. Then check and mark the memory accesses which are
+/// part of reduction like chains.
+///
+/// NOTE: This assumes independent blocks and breaks otherwise.
```
[the remark this answers is not in the document]
Comment (kartcq) [reply polly/lib/Analysis/ScopBuilder.cpp]: In line 2704, we check for all the binary instructions using this UsedOutsideBlock function as well.

**Comment by kartcq** [comment]
Ping @Meinersbur @efriedma-quic 
Note : Couple of changes are added after requesting re-review.
For latest changes build is clean.

**Comment by kartcq** [comment]
Ping @Meinersbur

**Comment by kartcq** [comment]
Ping @Meinersbur

**Comment by kartcq** [comment]
Ping @efriedma-quic @Meinersbur

File: polly/test/DependenceInfo/reduction_indirect_access.ll [hunk 6]
```diff
@@ -0,0 +1,39 @@
+; RUN: opt %loadPolly -basic-aa -polly-print-dependences -polly-allow-nonaffine -disable-output < %s | FileCheck %s
```
Comment (efriedma-quic) [inline polly/test/DependenceInfo/reduction_indirect_access.ll:1]: `-passes=print<polly-dependences>` (see #92918).  I think legacy-pm -polly-print-scops wasn't ported yet, so it should be okay to continue using it for now.

Otherwise, patch looks fine.

File: polly/lib/Analysis/ScopBuilder.cpp [hunk 7]
```diff
@@ -2557,7 +2568,6 @@ bool checkCandidatePairAccesses(MemoryAccess *LoadMA, MemoryAccess *StoreMA,
                      .intersect_domain(isl::manage(Domain.copy()));
     isl::set RS = R.range();
     isl::set WS = W.range();
-
```
Comment (Meinersbur) [inline polly/lib/Analysis/ScopBuilder.cpp:2560]: Please try to avoid unrelated changes.

**Review by Meinersbur — APPROVED** [review]
LGTM.

Please excuse the delay.

**Comment by kartcq** [comment]
Thanks @Meinersbur and @efriedma-quic for your time to review and approvals.
I have addressed all the final comments as well.
With the builders passing this patch is good to go for merging.

File: polly/test/DependenceInfo/reduction_indirect_access.ll (hunk 6 above)
Comment (efriedma-quic) [reply polly/test/DependenceInfo/reduction_indirect_access.ll:1]: Please don't forget to update this.

Comment (kartcq) [reply polly/test/DependenceInfo/reduction_indirect_access.ll]: Thanks for pointing out Eli. Updated.
